### ISSUE TITLE
sd-journal API facilitates creation of multiple mmap-cache instances in a single process, which isn't handled properly WRT SIGBUS

### ISSUE THREAD

**Issue by vcaputo** [systemd/systemd#21453]
As mentioned in https://github.com/systemd/systemd/pull/21452#issuecomment-974502640 , the current MMapCache implementation doesn't actually handle multiple MMapCache instances in the same process.

The sigbus processing consumes a global process-wide queue of sigbus faults via https://github.com/systemd/systemd/blob/f6278558da0304ec6b646bb172ce4688c7f162a5/src/basic/sigbus.c#L51 at https://github.com/systemd/systemd/blob/f6278558da0304ec6b646bb172ce4688c7f162a5/src/libsystemd/sd-journal/mmap-cache.c#L543 in service of a single, supplied MMapCache.

But sd-journal creates a new MMapCache instance for every journal_new() caller: sd_journal_open_namespace(), sd_journal_open_container(), sd_journal_open_directory(), sd_journal_open_files(), sd_journal_open_directory_fd(), sd_journal_open_files_fd().  There's nothing preventing a process from calling these multiple times...

So as-is we have a sigbus handling problem to deal with.  An obvious fix looks to be making mmap-cache maintain a global list of allocated MMapCache instances, and making mmap_cache_process_sigbus() iterate that list's MMapFileDescriptors before throwing its hands up and aborting.

I haven't chased down what the Golang bindings are doing here, but knowing Golang's tendency to mix "OS threads" and goroutines with reckless abandon, I suspect we also already have at least a theoretical problem WRT thread-safety and mmap_cache_prcess_sigbus() - especially once it's changed to iterate across all outstanding MMapCache instances process-wide.  I'd assume some big honking lock serializing at least all the MMapCaches for sigbus processing would be needed.

Worth noting: if the threaded case gets dealt with correctly for the sake of sd-journal, then it becomes kind of trivial to implement the isolated MMapFileDescriptor approach to threaded archiving via mmap-cache in journald.

**Comment by DaanDeMeyer** [comment]
I don't think this is a problem since sd-journal doesn't install our SIGBUS signal handler so SIGBUS signals won't ever get pushed onto the sigbus stack in the first place. Most sd-journal users will just coredump on SIGBUS unless they've installed their own handler.

**Comment by vcaputo** [comment]
That's a good point, I'd assumed the signal handler was being installed in all scenarios.

I wonder if we should just close this and leave it as a WONTFIX kind of thing, or if the nature of the bug is just different in that sd-journal isn't even facilitating proper handling of this despite it bringing mmap semantics into the calling process.

**Comment by DaanDeMeyer** [comment]
I don't think it would hurt to expose the sigbus logic that should be called from the SIGBUS signal handler so that users can call it from their own SIGBUS signal handler. (and maybe provide a function to install a SIGBUS signal handler as well since most users won't have an existing one).

We'd also have to fix sd-journal methods to add the `got_sigbus` logic that's currently only done in the write logic and not in the read logic.

(This is on top of the fixes you've proposed here already of course)
